From 91944709cf18c9a105521dbc2952c878895a8dfe Mon Sep 17 00:00:00 2001
From: Andreas Oberritter <obi@opendreambox.org>
Date: Wed, 20 Mar 2013 14:31:25 +0100
Subject: [PATCH 3/3] udhcpc: obtain hostname from OS by default

Doesn't require the 'hostname' option in /etc/network/interfaces, and thus
makes udhcpc behave like other clients in both (IPv4, IPv6) protocols.

Signed-off-by: Andreas Oberritter <obi@opendreambox.org>
Signed-off-by: Jens Rehsack <sno@netbsd.org>
---
 networking/udhcp/d6_dhcpc.c | 12 ++++++++++++
 networking/udhcp/dhcpc.c    | 12 ++++++++++++
 2 files changed, 24 insertions(+)

diff --git a/networking/udhcp/d6_dhcpc.c b/networking/udhcp/d6_dhcpc.c
index 85c410a7c..ad79d4ec0 100644
--- a/networking/udhcp/d6_dhcpc.c
+++ b/networking/udhcp/d6_dhcpc.c
@@ -1337,6 +1337,18 @@ int udhcpc6_main(int argc UNUSED_PARAM, char **argv)
 		logmode |= LOGMODE_SYSLOG;
 	}
 
+	/* try to get hostname from OS */
+	if (!client_data.hostname) {
+		char *hostname = safe_gethostname();
+		/*
+		 * Tito's safe_gethostname is finally a "safe_visible_gethostname",
+		 * but that's the most seen use-case.
+		 */
+		if (hostname[0] != '?')
+			client_data.hostname = alloc_dhcp_option(DHCP_HOST_NAME, hostname, 0);
+		free(hostname);
+	}
+
 	/* Create pidfile */
 	write_pidfile(client_data.pidfile);
 	/* Goes to stdout (unless NOMMU) and possibly syslog */
diff --git a/networking/udhcp/dhcpc.c b/networking/udhcp/dhcpc.c
index be978a05e..f083403f2 100644
--- a/networking/udhcp/dhcpc.c
+++ b/networking/udhcp/dhcpc.c
@@ -1411,6 +1411,18 @@ int udhcpc_main(int argc UNUSED_PARAM, char **argv)
 		logmode |= LOGMODE_SYSLOG;
 	}
 
+	/* try to get hostname from OS */
+	if (!client_data.hostname) {
+		char *hostname = safe_gethostname();
+		/*
+		 * Tito's safe_gethostname is finally a "safe_visible_gethostname",
+		 * but that's the most seen use-case.
+		 */
+		if (hostname[0] != '?')
+			client_data.hostname = alloc_dhcp_option(DHCP_HOST_NAME, hostname, 0);
+		free(hostname);
+	}
+
 	/* Create pidfile */
 	write_pidfile(client_data.pidfile);
 	/* Goes to stdout (unless NOMMU) and possibly syslog */
-- 
2.21.0 (Apple Git-122.2)

